test/boost/repair.cc

```diff
@@ -0,0 +1,137 @@
+#include "gc_clock.hh"
+#include "sstables/sstable_writer.hh"
+#include "readers/upgrading_consumer.hh"
+#include "readers/from_fragments_v2.hh"
+#include "test/lib/random_utils.hh"
+
+class test_mutation_fragment_queue_impl : public mutation_fragment_queue::impl, enable_lw_shared_from_this<test_mutation_fragment_queue_impl>{
+    schema_ptr _schema;
+    reader_permit _permit;
+    std::deque<mutation_fragment_v2> _fragments;
+    std::unique_ptr<upgrading_consumer<std::function<void(mutation_fragment_v2&&)>>> _consumer;
+    shared_sstable _sst;
+    sstable_writer_config _cfg;
+public:
+    test_mutation_fragment_queue_impl(schema_ptr s, reader_permit permit
+            , shared_sstable sst
+            , sstable_writer_config cfg)
+        : mutation_fragment_queue::impl()
+        , _schema(std::move(s))
+        , _permit(std::move(permit))
+        , _sst(std::move(sst))
+        , _cfg(std::move(cfg))
+    {
+        _consumer = std::make_unique<upgrading_consumer<std::function<void(mutation_fragment_v2&&)>>>
+            (*_schema, _permit, [this](mutation_fragment_v2&& mf) mutable {
+                _fragments.push_back(std::move(mf));
+            });
+    }
+    
+    virtual future<> push(mutation_fragment mf) override {
+        _consumer->consume(std::move(mf));
+        return make_ready_future();
+    }
+
+    virtual void abort(std::exception_ptr ep) override {}
+
+    virtual void push_end_of_stream() override {
+        flat_mutation_reader_v2 r = make_flat_mutation_reader_from_fragments(_schema, _permit, std::move(_fragments));
+        int estimated_partitions = 1;
+        _sst->write_components(std::move(r), estimated_partitions, _schema, _cfg, encoding_stats{}).get();
+        _sst->load().get();
+    }
+};
+
+mutation_fragment_queue make_test_mutation_fragment_queue(schema_ptr s, reader_permit permit, shared_sstable sst, sstable_writer_config cfg) {
+    return mutation_fragment_queue(std::make_unique<test_mutation_fragment_queue_impl>(std::move(s), std::move(permit), std::move(sst), std::move(cfg)));
+}
+
+class test_repair_writer_impl : public repair_writer::impl {
+    mutation_fragment_queue _queue;
+public:
+    test_repair_writer_impl(mutation_fragment_queue queue)
+        : _queue(std::move(queue))
+    {
+    }
+
+    virtual future<> wait_for_writer_done() override {
+        // Pushing EOS to the underlying mq waits until the sstable is written
+        return make_ready_future();
+    }
+
+    virtual mutation_fragment_queue& queue() override {
+        return _queue;
+    }
+
+    virtual void start_streaming(lw_shared_ptr<repair_writer> writer) override {}
+};
+
+lw_shared_ptr<repair_writer> make_test_repair_writer(schema_ptr schema, reader_permit permit, mutation_fragment_queue mq) {
+    return make_lw_shared<repair_writer>(std::move(schema), std::move(permit), std::make_unique<test_repair_writer_impl>(std::move(mq)));
+}
+
+SEASTAR_TEST_CASE(flush_repair_rows_on_wire_to_sstable) {
```

Comment:
Why flush to disk? Can't we compare two readers without going through an sstable? Asking in the interest of test speed.

Reply:
This is why it's important to add a comment on the test stating exactly which function or class or whatever it acts as a unit test for. If we're testing a function whose purpose is to write mutations to sstables, you can't test it without allowing this function to write the sstables. But if the tests is supposed to test repair's function which compares two mutation streams (and if I understand correctly, this is NOT what this test is about) then you could write the test for it without involving sstables.

Something I didn't understand, by the way, is why does the repair code have special code which writes mutations to sstables. Don't we have other code which writes mutations to sstables - and why would we want to test such code as part of repair's unit tests?

Reply:
What kind of special code do you mean? Repair uses the standard tools we have for writing a fragment stream into an sstable. The only special code it has is the multishard writer, which makes sure each partitions gets written to the shard which owns it. Repair runs on a single shard on each node, incoming data will usually cover any or all shards.

Reply:
I guess I just lost track of what exactly this unit-test actually tests. Not the full list of what the test **does** (which describes a lot of code included in the test itself, not in Scylla), but the actual Scylla function that is being **tested**. I thought that some sort of sstable-writer was being tested, otherwise I share Avi's question of why do we need to write an sstable.
I guess I'm just not familiar enough with this code to understand what it's doing without further explanations :-(

Reply:
I'll add the explanatory comment to the code, but in the interest of time and clarity let me write it here to get the feedback before I make another push of the patch series.

The test exercises a simplified path of what's happening during regular repair. In normal repair we receive repair_rows_on_wire, they go through various buffers and transformations and at some point are flushed to the disk. There is a lot of work related to figure out which rows to stream, where to pull them from etc. and that code is NOT part of this test.

The preparatory commits extract functions from repair meta that:
- take repair_rows_on_wire and convert them to std::list<repair_row> (repair_meta::to_repair_rows_lsit)
- flush std::list<repair_row> to disk (part of repair_meta::flush_rows_in_working_row_buf)
The test itself effectively just calls those two functions. The fluff around is only to randomly generate inputs and verify that whatever was written to disk is the same as the inputs.

The reason for preparatory work and a somewhat custom writer in the test is that the regular repair uses streaming::make_streaming_consumer, which requires (among other things) sharded<db::system_distributed_keyspace>. That doesn't make much sense in a single node unit test. That "somewhat custom writer" combines 3 functions used widely all over scylla so it's nothing fancy.

I'll reorganise the test code and write comments to make it clearer what the intention is.

With that preamble - I don't think we have to write to disk, though that's what the regular code does. I can rearrange the code to return the flat_mutation_reader_v2 from test_mutation_fragment_queue_impl::push_end_of_stream so that the test compares that reader with the memtable one (which is the source of mutations).

@avikivity As for speed of the test, I think it's acceptable in dev, I haven't checked debug, but will run it shortly.
➜  scylla git:(add-repair-unit-tests/v1) ✗ time ./test.py --mode=dev boost/repair_test
Found 1 tests.
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[1/1]       boost     dev   [ PASS ] repair_test.1
------------------------------------------------------------------------------
./test.py --mode=dev boost/repair_test  0.36s user 0.06s system 70% cpu 0.590 total

Reply:
Note that we have `cql_env`, which sets up a complete single node environment, including all the system tables. If you need it, feel free to use it. No need to make things more contrived, just to make the unit test a little simpler. We use it in a lot of unit test, when testing code which requires a working database, storage proxy, system tables, etc.

Reply:
I actually started with that, but it doesn't provide _system_distributed_keyspace, needed to create the streaming consumer that writes the mutations to sstables. After talking with @bhalevy, he mentioned that until we have multi-node test environments it doesn't make sense to have that distributed keyspace in cql_test_env (unless I'm misquoting). Maybe I could refactor and abstract away the usage of the keyspace in the consumer, but that sounded like avoiding one rabbit hole by going head first into another one.

Reply:
I think there is nothing preventing us from adding distributed system keyspace to `cql_test_env`, just like we have it set-up on single-node clusters. Don't know how trivial that is though.

Reply:
We actually already have it: https://github.com/scylladb/scylla/blob/59d56a3fd7ab22a3184de2ebfb3df67580316ff9/test/lib/cql_test_env.cc#L716
It is just not exposed directly.

Reply:
Reorganised the test code, added comments and removed flushing in favour of comparing memtable reader + reader from fragments that would have been passed to sstable::write_components.